### PR TITLE
allow batch enqueueing of jobs using Redis Pipelining

### PR DESCRIPTION
**Pull request by jack-kerouac**
This change allows enqueueing a list of jobs with higher performance than one-by-one invoking the existing `client.enqueue()` method. It therefore uses Redis Pipelining: https://redis.io/topics/pipelining.

I decided against supporting splitting the list of jobs into chunks of e.g. 10,000 jobs, which is recommended by Redis ('Important Note' at https://redis.io/topics/pipelining#redis-pipelining). I mentioned this in the JavaDoc and left it to the client of this method to keep the library simple. Some clients might want to ignore this and I wanted to leave it up to the user.

This change also includes a bit of refactoring of the `validate` methods to be able to validate the queue name parameter independently of a job parameter.

Closes https://github.com/gresrun/jesque/issues/123

### CODE REVIEW

**Comment by jack-kerouac** [comment]
@maciejp: On top of yesterday's work, I added some more JavaDocs, created an `ArrayList` with given capacity, mentioned the change in the README, and refactored some of the validate methods.

This is a PR to our own master branch. But I would when you are fine with this open the exact same PR to the original library.

File: src/main/java/net/greghaines/jesque/client/AbstractClient.java [hunk 1]
```diff
@@ -38,7 +42,7 @@
 
     /**
      * Constructor.
-     *
```
Comment (maciejp) [inline src/main/java/net/greghaines/jesque/client/AbstractClient.java]: unnecessary change (in a few places below as well)

File: src/main/java/net/greghaines/jesque/client/AbstractClient.java [hunk 2]
```diff
@@ -132,9 +158,21 @@ public boolean acquireLock(final String lockName, final String lockHolder, final
...
+     * Actually enqueue the serialized jobs.
+     *
+     * @param queue
+     *            the queue to add the Jobs to
+     * @param msg
+     *            the serialized Jobs
+     * @throws Exception
+     *             in case something goes wrong
+     */
+    protected abstract void doBatchEnqueue(String queue, List<String> msg) throws Exception;
```
Comment (maciejp) [inline src/main/java/net/greghaines/jesque/client/AbstractClient.java]: `msg` -> `msgs` ? (plural)

File: src/main/java/net/greghaines/jesque/client/AbstractClient.java [hunk 3]
```diff
@@ -157,13 +195,13 @@ public boolean acquireLock(final String lockName, final String lockHolder, final
      * @throws Exception
      *             in case something goes wrong
      */
-    protected abstract boolean doAcquireLock(final String lockName, final String lockHolder, 
+    protected abstract boolean doAcquireLock(final String lockName, final String lockHolder,
```
Comment (maciejp) [inline src/main/java/net/greghaines/jesque/client/AbstractClient.java]: unnecessary change

File: src/main/java/net/greghaines/jesque/client/Client.java [hunk 4]
```diff
@@ -37,9 +39,23 @@
...
+     * Queues jobs in a given queue to be run. It uses Redis Pipelining (https://redis.io/topics/pipelining).
+     * Check out the "Important Note" here: https://redis.io/topics/pipelining#redis-pipelining. => Consider splitting
+     * long lists of jobs into chunks of 10,000 or so.
+     *
+     * @param queue
+     *            the queue to add the Job to
+     * @param jobs
+     *            the jobs to be enqueued
+     * @throws IllegalArgumentException
+     *             if the queue is null or empty or if the list of jobs is null
```
Comment (maciejp) [inline src/main/java/net/greghaines/jesque/client/Client.java:52]: looks to me that passing null as a list of jobs would result in a NPE

File: src/test/java/net/greghaines/jesque/IntegrationTest.java [hunk 5]
```diff
@@ -217,6 +229,24 @@ private static void assertMixed(final WorkerListener listener, final WorkerEvent
...
+        final Jedis jedis = createJedis(CONFIG);
+        try {
+            Assert.assertEquals("2", jedis.get(createKey(CONFIG.getNamespace(), STAT, FAILED)));
+            Assert.assertEquals("2", jedis.get(createKey(CONFIG.getNamespace(), STAT, PROCESSED)));
+        } finally {
+            jedis.quit();
+        }
+    }
+
     private static void doWork(final List<Job> jobs, final Map<String, ? extends Class<? extends Runnable>> jobTypes,
```
Comment (maciejp) [inline src/test/java/net/greghaines/jesque/IntegrationTest.java:250]: one small refactoring that we could do is changing the signature of this method & TestUtils.enqueueJobs to accept a single job and then use it wherever the list of jobs has 1 element.

worker creation could also be extracted to a method

**Review by maciejp — CHANGES_REQUESTED** [review]
all good, just one minor thing with the NPE if I'm not mistaken.
the refactoring I mentioned is purely optional and I'm actually not sure if modifying the test behaviour (replacing the one-by-one enqueuing with batch enqueuing for all lists) is a good idea.

File: README.md [hunk 6]
```diff
@@ -54,6 +54,9 @@ worker.end(true);
 try { workerThread.join(); } catch (Exception e){ e.printStackTrace(); }
 ```
 
+If enqueueing multiple jobs at the same time, there is `client.batchEnqueue(String queue, List<Job> jobs)` which does it
```
Comment (maciejp) [inline README.md:57]: should we maybe mention how does it do it and why is it more optimal?

Comment (jack-kerouac) [reply README.md:57]: I think this is too detailed for the README. The section is called 'Quickstart'.

File: src/main/java/net/greghaines/jesque/client/Client.java (hunk 4 above)
Comment (jack-kerouac) [reply src/main/java/net/greghaines/jesque/client/Client.java:52]: Correct. Fixed that.

File: src/test/java/net/greghaines/jesque/IntegrationTest.java (hunk 5 above)
Comment (jack-kerouac) [reply src/test/java/net/greghaines/jesque/IntegrationTest.java:250]: I would not do this right now. Let's keep it as it is.

**Review by maciejp — APPROVED** [review]

**Comment by jack-kerouac** [comment]
Will close this PR in favor of opening one to the upstream repo.
